### ISSUE TITLE
setup.rb does not respec --destdir for Bundler

### ISSUE THREAD

**Issue by voxik** [rubygems/rubygems#3604]
Trying to package updated version of RubyGems, the setup.rb command fails:

~~~
$ ruby setup.rb --document rdoc,ri --prefix=/ --backtrace --no-regenerate-binstubs --destdir=/builddir/build/BUILDROOT/rubygems-3.1.3-200.fc33.x86_64/usr/share/rubygems/
ERROR:  While executing gem ... (Errno::EACCES)
    Permission denied @ apply2files - /usr/share/gems/specifications/bundler-2.1.4.gemspec
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/commands/setup_command.rb:408:in `delete'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/commands/setup_command.rb:408:in `install_default_bundler_gem'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/commands/setup_command.rb:174:in `execute'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/command.rb:325:in `invoke_with_build_args'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/command_manager.rb:178:in `process_args'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/command_manager.rb:148:in `run'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/gem_runner.rb:59:in `run'
	setup.rb:41:in `<main>'
~~~

Obviously, it tries to touch system version of the Bundler, not respecting the `--destdir` option.

It seems it was not problem previously (last time I tried with 3.0.3), because Bundler was installed just conditionally prior f9791cae3aa99e53e735.

**Comment by deivid-rodriguez** [comment]
Is this the same as https://github.com/rubygems/rubygems/pull/2768?

**Comment by voxik** [comment]
I have fixed the first issue by this patch:

~~~
$ git diff
diff --git a/lib/rubygems/commands/setup_command.rb b/lib/rubygems/commands/setup_command.rb
index d67caca9..e0ab3154 100644
--- a/lib/rubygems/commands/setup_command.rb
+++ b/lib/rubygems/commands/setup_command.rb
@@ -411,6 +411,7 @@ By default, this RubyGems will install gem as:
 
     # Remove gemspec that was same version of vendored bundler.
     normal_gemspec = File.join(Gem.default_dir, "specifications", "bundler-#{bundler_spec.version}.gemspec")
+    normal_gemspec = File.join(options[:destdir], normal_gemspec) unless Gem.win_platform?
     if File.file? normal_gemspec
       File.delete normal_gemspec
     end
~~~

But there is next one right away:

~~~
ruby setup.rb --document rdoc,ri --prefix=/ --backtrace --no-regenerate-binstubs --destdir=/builddir/build/BUILDROOT/rubygems-3.1.3-200.fc33.x86_64/usr/share/rubygems/
ERROR:  While executing gem ... (Errno::EACCES)
    Permission denied @ rb_sysopen - /usr/share/gems/specifications/default/bundler-2.1.4.gemspec
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/installer.rb:471:in `initialize'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/installer.rb:471:in `open'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/installer.rb:471:in `write_default_spec'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/installer.rb:323:in `install'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/commands/setup_command.rb:432:in `block in install_default_bundler_gem'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/commands/setup_command.rb:428:in `chdir'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/commands/setup_command.rb:428:in `install_default_bundler_gem'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/commands/setup_command.rb:174:in `execute'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/command.rb:325:in `invoke_with_build_args'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/command_manager.rb:178:in `process_args'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/command_manager.rb:148:in `run'
	/builddir/build/BUILD/rubygems-3.1.3/lib/rubygems/gem_runner.rb:59:in `run'
	setup.rb:41:in `<main>'
~~~

Interestingly, the exception is handled somewhere and the setup succeeds

**Comment by voxik** [comment]
> Is this the same as #2768?

I don't think so.

**Comment by voxik** [comment]
This line:

https://github.com/rubygems/rubygems/blob/8d54023b0ba57c4c6521678bd3426576aae1f317/lib/rubygems/installer.rb#L440

is to blame for the error above. It should somehow prepend the `options[:build_root]`. But I can't see any reasonable way forward :/ The `default_spec_file` should probably not be used in `write_default_spec` and something like `spec.spec_dir + 'default' + spec.file_name` should be used instead.

**Comment by deivid-rodriguez** [comment]
@voxik I tried your example and I didn't get the same errors you were getting when you reported this issue, everything seemed to be installed under `destdir`. I did run into two issues:

* `--destdir` was getting duplicated, so if you passed `tmp/foo`, some stuff would get installed under `/tmp/foo/tmp/foo`. That should've been fixed by #2768.
* Sometimes `--prefix` would get ignored and the default prefix would be prepended to some locations. I expect #5051 to fix that.

Can you verify whether it all works now with #5051? I think there are still issues with `--regenerate-plugins` and `--regenerate-binstubs`, but I'll get to those later.

**Comment by deivid-rodriguez** [comment]
Also created https://github.com/rubygems/rubygems/pull/5052 to fix binstub and plugin regeneration. That PR should contain all required changes, so I suggest to try that.

**Comment by deivid-rodriguez** [comment]
Found one more place where we were trying to write outside of `--destdir` thanks to the report in #4508, and that should be fixed by #5053. I think `--destdir` and `--prefix` should work just fine now.
